### PR TITLE
Fix mask slicing for models with HybridCache

### PR DESCRIPTION
**Pull request by Cyrilvallez**
# What does this PR do?

As per the title. Models with HybridCache need to correctly slice the key/value states when using FA2 as inputs needs to be unpadded on the right as well (and the mask has shape [bs, seq_len]). Moreover, mask slicing was wrong in all cases when the sequence length is larger than the sliding window. It is currently broken and leads to garbage generation when using padding. This fixes it.

### CODE REVIEW

File: src/transformers/models/gemma2/modeling_gemma2.py [hunk 1]
```diff
@@ -278,18 +283,13 @@ def forward(
...
-            else:
+            # For FA2, the mask is 2D and is of shape [bs, seq_len] (not [bs, fixed_cache_len])
+            if self.config._attn_implementation != "flash_attention_2":
                 min_dtype = torch.finfo(hidden_states.dtype).min
                 sliding_window_mask = torch.tril(
                     torch.ones_like(attention_mask, dtype=torch.bool), diagonal=-self.sliding_window
                 )
                 attention_mask = torch.where(sliding_window_mask, min_dtype, attention_mask)
-                if attention_mask.shape[-1] <= 1:  # when decoding
-                    attention_mask = attention_mask[:, :, :, -self.sliding_window :]
```
Comment (ArthurZucker) [inline src/transformers/models/gemma2/modeling_gemma2.py:292]: this is super important ! Why is it removed? 
I know it is counter intuitive, but `_flash_attention_forward` takes the attention mask to pad / unpad the input itds. 
Thus you need the slicing otherwise this operation fails, see the blame !

**Review by ArthurZucker — COMMENTED** [review]
For gemma2 it was suppose to work!
Not aligned with the removal of the attentoin mask slicing tho! Let's run slow test on the PR

File: src/transformers/models/gemma2/modeling_gemma2.py (hunk 1 above)
Comment (Cyrilvallez) [reply src/transformers/models/gemma2/modeling_gemma2.py:292]: Indeed, I was too fast on this one, the HybridCache behaves slightly differently than I remembered. There was still an issue in the slicing during prefill for FA2 though!

**Comment by Cyrilvallez** [comment]
Ok, should now be all good @ArthurZucker!
Confirming that slow tests are the same as on main as well

**Comment by Cyrilvallez** [comment]
PS: slow tests are similar as on main for both models (and the new tests showing full equivalence all pass)

File: src/transformers/models/cohere2/modular_cohere2.py [hunk 2]
```diff
@@ -473,9 +489,20 @@ def forward(
...
 
+        # This is needed to correctly slice the mask without data-dependent slicing later on if using dynamo tracing
+        # (retrieving the same value from `cache_position` later on would crash dynamo)
+        if last_cache_position is None:
+            last_cache_position = 0
+            if attention_mask is not None:
+                # In case a 4d mask is passed directly without using `generate`, we have to rely on cache_position
+                # It will break dynamo tracing but there are no way around it (and it should never happen in practice)
+                last_cache_position = (
+                    attention_mask.shape[-1] if attention_mask.dim() == 2 else cache_position[-1].item()
```
Comment (ArthurZucker) [inline src/transformers/models/cohere2/modular_cohere2.py:500]: if in practice it never happens, let's remove this one

**Review by ArthurZucker — COMMENTED** [review]
Okay, I understand the motivation, and as we have kwargs I'd be in favor of passing them through kwargs but explicit is also fine. 

Wondering wether we want to do a lot of work for the user or not. I am also thinking a bit more with continuous batching were you want to prepare 2 different tensors: you have the index to fill and the cache index to slice, so will be a bit aligned with this. 

TLDR; we do need to patch Gemma2 and cohere, so let's iterate a bit! Only bothered by the case that won't happen in practice!

File: src/transformers/models/gemma2/modeling_gemma2.py [hunk 3]
```diff
@@ -592,6 +614,16 @@ def forward(
...
 
+        # This is needed to correctly slice the mask without data-dependent slicing later on if using dynamo tracing
+        # (retrieving the same value from `cache_position` later on would crash dynamo)
+        if last_cache_position is None:
+            last_cache_position = 0
+            if attention_mask is not None:
+                # In case a 4d mask is passed directly without using `generate`, we have to rely on cache_position
+                # It will break dynamo tracing but there are no way around it (and it should never happen in practice)
+                last_cache_position = (
+                    attention_mask.shape[-1] if attention_mask.dim() == 2 else cache_position[-1].item()
```
Comment (ArthurZucker) [inline src/transformers/models/gemma2/modeling_gemma2.py:625]: for gemma 2 I would say passing the mask of a correct shape might be on the users's side to chefck

**Review by ArthurZucker — APPROVED** [review]
Thanks IMO we should remove the 4d without generate case as the users is expected to have correct mask, unless you are in the decoder layer and then you slice depending on layers

**Comment by ani300** [comment]
Any updates on when this is merging? This fix is blocking https://github.com/vllm-project/vllm/pull/12120 and https://github.com/vllm-project/vllm/pull/10909
